codetf.json

```diff
@@ -37,6 +37,7 @@
                 "changes" : [
                     {
                         "lineNumber" : "153", // the "primary" line number that was changed (required)
+                        "diffSide": "right", // "right" or "left". "right" indicates a new or modified line of code. "left" indicates a line of code that was deleted (required)
```

Comment:
consider renaming this `side` to align with GitHub's API (see [create PR comment endpoint docs](https://docs.github.com/en/rest/pulls/comments?apiVersion=2022-11-28#create-a-review-comment-for-a-pull-request)). I don't think "alignment with GitHub API" is a strict goal, but if there's no reason not to, we may as well, to keep our code easier to read.

Reply:
I'm a little worried that by doing that we obscure the intent of this field since understanding its purpose would mean having some context about the underlying implementation of the consumer. I'd prefer to preserve as much clarity here as possible, although I'm certainly open to a better name than `diffSide`. That's what brought me originally to the idea of `changeType`.

What about something that makes its relationship to the line number more explicit? Like `lineNumberLocation` and the values are `"original"` and `"new"`?

tl;dr: naming things is hard.

Reply:
I understand, but my point is that CodeTF does not and should not know anything at all about the GitHub API. What is more compelling is to use a field that makes sense in the context of a CodeTF change. 

In the context of CodeTF what we're really trying to say is whether the given line number and description refer to the original code or whether they refer to new code. I think we should do our best to make that connection explicit when naming this field (and I admit that `diffSide` does no such thing) rather than coupling this field to a largely contingent and invisible implementation detail of our platform.

If you feel strongly that `side` is the most useful term here, then I would argue that it really belongs in the `properties` bucket in which case it can't be a required field.

Reply:
Fwiw, the change that would make me the happiest is to add `side` to the `properties` bucket since it is explicitly for such platform-specific purposes. However, I approached this with the assumption that this needs to be a required field, which necessitates a spec change.

If you're okay with using `properties` for this purpose, then I will close this PR and make the change to Python codemodder. This means we need to agre on a reasonable default. As a compromise I am happy to conform to the existing platform behavior and make Python explicit in its requirements. I think this would preclude the need for a lot of the other discussions on this thread since we can adapt to our needs without further spec changes.

The downside of this is that future codemodders are going to get the default behavior, which may not be what they want. But that's a problem for future you and me, and not necessarily one that needs to be specced.

Reply:
I think the `properties` bag should be reserved only for cases where we can't possibly justify including some data in the CodeTF spec proper. It's ripe for abuse and I kind of wish it didn't exist.

We'll just map whatever term you choose to `side`.

Reply:
There's a lot here, and I only want to state a few opinions and then get out of the way:

1. I agree that I would like CodeTF to be as neutral as possible to platform or SCM needs. I think `diffSide` is more clear to codemod authors, who we want to attract. The platform is a bit of a captured audience.
2. We have 2 codemodders right now and it is convenient for them to provide opposite `diffSide` values. I think there should be no default and we should force codemodder _developers_ to figure this out for themselves. If it's required, it's easy to provide. If it's not required,  they will likely be subtly wrong and maybe not notice.
3. This should not be a `properties` entry. It is a part of storytelling, which is a top concern and differentiator for codemodder.

Reply:
What about `diffSide` with `"new"` and `"original"` to make the connection more explicit?

Reply:
Just want to point out that these numbers could be wrong in the situation where we've made multiple changes to the same file across multiple codemods, right? The first line number will be correct, but any subsequent changes will be wrong. I think aiming for perfection here isn't possible, so I'd be nervous to keep adding complexity here.

We may eventually be able to add AI more appropriately guess the best location for these?

Reply:
In principle, yes. In practice, it is incumbent upon the user (e.g. the platform) to properly account for this before invoking codemodder and rendering any results.

Reply:
Wait, why would the line numbers necessarily be wrong? Each codemod's diff builds on the previous codemod's diff, so why would this be different for the line numbers in our storytelling?

Reply:
Codemod A inserts a line at line 10. It records the change at line 10.

Codemod B inserts at line 5. It records the change at line 5.

Codemod A change was recorded at the correct "left" location after the next codemod executed, the recorded left" location will be wrong.

Reply:
I see. I think it's correct to say that if the platform were commenting on _each commit's diff_ then the lines would be correct, but the platform is actually creating _PR comments_, and PRs mash-up many commits into one diff view.